sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Cast.scala

```diff
@@ -456,9 +487,54 @@ case class Cast(child: Expression, dataType: DataType, timeZoneId: Option[String
     case DateType =>
       buildCast[Int](_, d => null)
     case TimestampType =>
-      buildCast[Long](_, t => timestampToLong(t).toInt)
-    case x: NumericType =>
-      b => x.numeric.asInstanceOf[Numeric[Any]].toInt(b)
+      buildCast[Long](_, t => {
+        val longValue = timestampToLong(t)
+        if (longValue == longValue.toInt) {
+          longValue.toInt
+        } else {
+          null
+        }
+      })
+    case ByteType =>
+      b => b.asInstanceOf[Byte].toInt
+    case ShortType =>
+      b => b.asInstanceOf[Short].toInt
+    case LongType =>
+      buildCast[Long](_, l =>
+        if (l == l.toInt) {
+          l.toInt
+        } else {
+          null
+        }
+      )
+    case FloatType =>
+      buildCast[Float](_, f =>
+        if (f <= Int.MaxValue && f >= Int.MinValue) {
```

Comment:
Why don't we use the similar condition to others, like `if (f < Int.MaxValue + 1L && f > Int.MinValue - 1L) {`?

Reply:
It is quite tricky here. I did some corner tests and decide to do it in this way.  
```
scala> 2147483647.5f < Int.MaxValue + 1L
res2: Boolean = false

scala> 2147483647.5f <= Int.MaxValue 
res3: Boolean = true
```